overlord/fdestate/fdemgr.go

```diff
@@ -412,6 +414,55 @@ func GetRecoveryKey(st *state.State, keyID string) (rkey keys.RecoveryKey, err e
 	return mgr.getRecoveryKey(keyID)
 }
 
+// devPathByContainerRole returned a map from container role to corresponding
+// device path. If containerRoles is nil, all encrypted containers are returned.
+// If any encrypted container role from containerRoles does not exist, an error
+// is returned.
+func (m *FDEManager) devPathByContainerRole(containerRoles []string) (map[string]string, error) {
+	containers, err := m.GetEncryptedContainers()
+	if err != nil {
+		return nil, err
+	}
+
+	allContainers := len(containerRoles) == 0
+	devPathByContainerRole := make(map[string]string, len(containers))
+	for _, container := range containers {
+		if allContainers || strutil.ListContains(containerRoles, container.ContainerRole()) {
+			devPathByContainerRole[container.ContainerRole()] = container.DevPath()
```

Comment:
Should we warn or something if multiple devices have the same container role?

Reply:
This feels more like something to do in the Ensure loop?